{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7b0eeff3-68f4-47d8-84f1-56b3c83e1be7",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Documentation on running the TRIMBA and RNASEQ pipelines"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "430e6d09-124f-4c78-8d71-6d505ae9a829",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Test your Ubuntu installation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "911fdc10-6aad-4e7d-851a-4283d87de8fd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "To install the pipeline on your local Ubuntu machine, ensure you have `git`, `docker` and `nextflow` installed.\n",
    "\n",
    "To check if you have `git` execute the following commands on the Ubuntu command line.\n",
    "\n",
    "```bash\n",
    "git --version\n",
    "```\n",
    "\n",
    "That should say something like:\n",
    "```\n",
    "git version 2.34.1\n",
    "```\n",
    "\n",
    "If not, **you need to install git**.\n",
    "\n",
    "```bash\n",
    "sudo apt-get install git\n",
    "```\n",
    "\n",
    "### Test if you have docker access\n",
    "\n",
    "The pipeline is using `docker` and to test this is working on your computer, try to following command:\n",
    "\n",
    "```bash\n",
    "docker run hello-world\n",
    "```\n",
    "\n",
    "That should give something like:\n",
    "\n",
    "```\n",
    "\n",
    "Hello from Docker!\n",
    "This message shows that your installation appears to be working correctly.\n",
    "\n",
    "To generate this message, Docker took the following steps:\n",
    " 1. The Docker client contacted the Docker daemon.\n",
    " 2. The Docker daemon pulled the \"hello-world\" image from the Docker Hub.\n",
    "    (amd64)\n",
    " 3. The Docker daemon created a new container from that image which runs the\n",
    "    executable that produces the output you are currently reading.\n",
    " 4. The Docker daemon streamed that output to the Docker client, which sent it\n",
    "    to your terminal.\n",
    "\n",
    "To try something more ambitious, you can run an Ubuntu container with:\n",
    " $ docker run -it ubuntu bash\n",
    "\n",
    "Share images, automate workflows, and more with a free Docker ID:\n",
    " https://hub.docker.com/\n",
    "\n",
    "For more examples and ideas, visit:\n",
    " https://docs.docker.com/get-started/\n",
    " ```\n",
    " \n",
    " If docker is not installed, check the following on how to install docker on Windows machine and configure it for use on your Ubuntu WSL2 (Windows Subsystem for Linux, version 2) system.\n",
    "\n",
    "https://docs.docker.com/desktop/wsl/\n",
    "\n",
    "#### Permission denied error\n",
    "\n",
    "When you get the following when trying to run `docker run hello-world`:\n",
    "\n",
    "```\n",
    "\"Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post \"http://%2Fvar%2Frun%2Fdocker.sock/v1.24/containers/create\": dial unix /var/run/docker.sock: connect: permission denied.\n",
    "```\n",
    "\n",
    "it means that the uesr you run at, is not yet a member of to `docker` group. Either run as `sudo` or better, add the user to the `docker` group with the following command:\n",
    "\n",
    "```bash\n",
    "sudo usermod -aG docker $(whoami)\n",
    "```\n",
    "\n",
    "Now you should be able to run:\n",
    "\n",
    "```bash\n",
    "docker run hello-world\n",
    "```\n",
    "\n",
    "and get something like this:\n",
    "```\n",
    "Hello from Docker!\n",
    "This message shows that your installation appears to be working correctly.\n",
    "\n",
    "To generate this message, Docker took the following steps:\n",
    " 1. The Docker client contacted the Docker daemon.\n",
    " 2. The Docker daemon pulled the \"hello-world\" image from the Docker Hub.\n",
    "    (amd64)\n",
    " 3. The Docker daemon created a new container from that image which runs the\n",
    "    executable that produces the output you are currently reading.\n",
    " 4. The Docker daemon streamed that output to the Docker client, which sent it\n",
    "    to your terminal.\n",
    "\n",
    "To try something more ambitious, you can run an Ubuntu container with:\n",
    " $ docker run -it ubuntu bash\n",
    "\n",
    "Share images, automate workflows, and more with a free Docker ID:\n",
    " https://hub.docker.com/\n",
    "\n",
    "For more examples and ideas, visit:\n",
    " https://docs.docker.com/get-started/\n",
    "```\n",
    "\n",
    "### Test if you have nextflow\n",
    "\n",
    "The pipelines are executed with [`nextflow`](https://www.nextflow.io/) and you can check if `nextflow` is installed with this command:\n",
    "\n",
    "```\n",
    "nextflow -v\n",
    "```\n",
    "\n",
    "Which should give something like:\n",
    "\n",
    "```\n",
    "nextflow version 23.10.1.5891\n",
    "```\n",
    "If it does not, you need to install it. `nextflow` has one more requirement, which is `java`, a programming language that is used by `nextflow` so ensure you have `java` by running this command:\n",
    "\n",
    "```bash\n",
    "java -version\n",
    "```\n",
    "\n",
    "that should give something like:\n",
    "```\n",
    "openjdk version \"11.0.21\" 2023-10-17\n",
    "OpenJDK Runtime Environment (build 11.0.21+9-post-Ubuntu-0ubuntu122.04)\n",
    "OpenJDK 64-Bit Server VM (build 11.0.21+9-post-Ubuntu-0ubuntu122.04, mixed mode, sharing)\n",
    "```\n",
    "\n",
    "**Note** It is important that it is version 11 (11.0.21 in my case). If this is smaller than 11, you need to install JAVA 11. More information on installing `java 11` you can find here: https://www.nextflow.io/docs/latest/getstarted.html#requirements\n",
    "\n",
    "If `java 11` is installed, you can use the following command to install nextflow:\n",
    "\n",
    "```bash\n",
    "curl -s https://get.nextflow.io | bash\n",
    "```\n",
    "\n",
    "This will install everything you need and place it likely in your home directory in `.local/bin/nextflow`.\n",
    "\n",
    "If, after installation the command `nextflow -v` does still not work, type this command:\n",
    "\n",
    "```bash\n",
    "${HOME}/.local/bin/nextflow -v\n",
    "```\n",
    "\n",
    "If that works, it means that `~/.local/bin` is not on your `PATH` and you can add that to your path in your `.bashrc` file by adding the following as the last line in that file\n",
    "\n",
    "```\n",
    "export PATH=${HOME}/.local/bin:$PATH\n",
    "```\n",
    "\n",
    "That completes the requirements check. Next is to install the `trimba` pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2de0c737-7be8-43df-8f0c-5cc845b70a3e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Install the pipeline locally"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d357bbd4-74a6-486d-a049-fe587f354a8e",
   "metadata": {
    "tags": []
   },
   "source": [
    "To get the `trimba` pipeline, you need to install it locally on the `ubuntu` system. First create or navigate to the folder you want to install it in and then execute the following command:\n",
    "\n",
    "```\n",
    "git clone https://github.com/thondeboer/nf-core-trimba.git\n",
    "```\n",
    "\n",
    "that will install folder that is called `nf-core-trimba` in the folder and that will be the value you use in the `nextflow` command."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472330f9-232c-43a3-a8cc-9f56394a7d30",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Do test run of the RNAseq pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024b241d-376a-450f-bf0a-9dc9299e010d",
   "metadata": {},
   "source": [
    "The RNASEQ pipeline does not need to be installed with `git` but will automatically be downloaded when you run it for the first time. It is a good practice to test your setup and it will download all the data it needs as well, so is very simple to run\n",
    "\n",
    "```bash\n",
    "nextflow run nf-core/rnaseq -r 3.14.0 -profile test,docker --outdir test_rnaseq\n",
    "```\n",
    "\n",
    "This will create a directory test_rnaseq and run the pipeline. Takes a few minutes (~4m on my computer) and should complete successfully."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a737b41-caa9-46c1-bf71-e3aca933b345",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Configure and run the `trimba` pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74b2d4d5-7467-4559-94bb-658ca0b844be",
   "metadata": {},
   "source": [
    "The `trimba` and `rnaseq` pipeline require some configurations, revolving mostly around providing the pipeline with the files to analyze.\n",
    "\n",
    "The script `run_trima_and_rnaseq.sh` contains the complete set of commands to run and you should edit this file for each run.\n",
    "\n",
    "The next section breaks down each of the sections in the pipeline script."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcd3f7e7-96fb-4841-95fa-3b8191d869cc",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Main configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86522462-1f51-40ea-90e7-834b728fe4a9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "Most of the confguration is done through using shell script variables and hopefully only a few need to be changed to reflect your situation and would only need to be changed ONCE.\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#####\n",
    "#\n",
    "# MAIN CONFIGURATION\n",
    "#\n",
    "\n",
    "# Location of your FASTQ data\n",
    "DATADIR='/mnt/data/yasar'\n",
    "FASTQ_DIR=\"${DATA_DIR}/FASTQ\"\n",
    "\n",
    "# The locatiom of your reference files (Genome FASTA and indices etc.)\n",
    "REF=\"/mnt/data/REFERENCES\"\n",
    "\n",
    "# The location of the PIPELINES\n",
    "PIPELINE_DIR=\"${HOME}/CODE/workflows\"\n",
    "TRIMBA=\"${PIPELINE_DIR}/nf-core-trimba\"\n",
    "\n",
    "# The location of the nextflow exectuable\n",
    "NEXTFLOW=\"${HOME}/.local/bin/nextflow\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54318911-89fa-4211-a328-2c6e0edd32bd",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Reference files configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63777db0-f38a-404d-9377-4f2526ecb327",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "The next section contains the configuration of all the various reference files that the pipeline needs. This is also a section that only need to be changed ONCE and then used for all pipeline runs.\n",
    "\n",
    "```bash\n",
    "#####\n",
    "#\n",
    "# REFERENCE FILES CONFIGURATION\n",
    "#\n",
    "\n",
    "# Mouse data - REQUIRED - Only these THREE files are required\n",
    "# See the documentation on how to download these files\n",
    "GRCM38_FASTA=\"${REF}/nf-core/GRCm38/fasta/default/Mus_musculus.GRCm38.dna.primary_assembly.fa\"\n",
    "GRCM38_GTF=\"${REF}/nf-core/GRCm38/ensembl_gtf/default/Mus_musculus.GRCm38.102.gtf.gz\"\n",
    "MEME_DB=\"${REF}/memesuite/motif_databases\"\n",
    "\n",
    "# Mouse data - OPTIONAL (If not provided, will be created at run time but with save_reference can be saved for next time)\n",
    "GRCM38_GTF_DB=\"${REF}/nf-core/GRCm38/ensembl_gtf/default/gtf.db\"\n",
    "GRCM38_BOWTIE_IDX=\"${REF}/nf-core/GRCm38/fasta/default/bowtie_index\"\n",
    "GRCM38_RNAS_BOWTIE_IDX=\"${REF}/nf-core/GRCm38/fasta/default/Mus_musculus.GRCm38.dna.primary_assembly.RNAtranscripts_bowtie\"\n",
    "GRCM38_TRANSCRIPTS=\"${REF}/nf-core/GRCm38/fasta/default/Mus_musculus.GRCm38.dna.primary_assembly.ALLtranscripts.fa\"\n",
    "GRCM38_TRANSCRIPTS_BOWTIE_IDX=\"${REF}/nf-core/GRCm38/fasta/default/Mus_musculus.GRCm38.dna.primary_assembly.ALLtranscripts_bowtie\"\n",
    "# For RNASEQ\n",
    "GRCM38_STAR_INDEX=\"${REF}/nf-core/GRCm38/fasta/default/star\"\n",
    "GRCM38_SALMON_INDEX=\"${REF}/nf-core/GRCm38/fasta/default/salmon\"\n",
    "\n",
    "#*****************************\n",
    "#* END OF MAIN CONFIGURATION *\n",
    "#*****************************\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29afa252-af2a-48db-8456-ef9208419d06",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Pipline configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11ace1f5-7e0e-4408-b4e9-7c3439dcef23",
   "metadata": {},
   "source": [
    "The next sction is the pipeline configuration and this is likely an area that needs to be configured for **each run of the pipelines**.\n",
    "\n",
    "There are six main configuration settings to review\n",
    "\n",
    "1. OUTDIR_TRIMBA      : The results directory for the `trimba` pipeline\n",
    "1. OUTDIR_RNASEQ      : The results directory for the `nf-core/rnaseq` pipeline.\n",
    "1. SAMPLESHEET_TRIMBA : The file containing the input files for the `trimba` pipeline.\n",
    "1. SAMPLESHEET_RNASEQ : The file containing the input files for the `nf-core/rnaseq` pipeline.\n",
    "1. GENES_OF_INTEREST  : The genes of interest for the `trimba` pipeline.\n",
    "1. MOTFIS_FILE        : The motifs of interest for the `trimba` pipeline.\n",
    "\n",
    "```bash\n",
    "#####\n",
    "#\n",
    "# OUTDIR CONFIGURATION - Edit this\n",
    "#\n",
    "# The OUTDIR determines where all the inputs and outputs are stored.\n",
    "OUTDIR_TRIMBA=\"${DATA_DIR}/WF/trimba_demo\"\n",
    "OUTDIR_RNASEQ=\"${DATA_DIR}/WF/rnaseq_demo\"\n",
    "\n",
    "#####\n",
    "#\n",
    "# FILE SETUP - Do not edit\n",
    "#\n",
    "# This section just defines some more variables and should not be edited\n",
    "mkdir -p $OUTDIR_TRIMBA && cd $OUTDIR_TRIMBA\n",
    "SAMPLESHEET_TRIMBA=\"${OUTDIR_TRIMBA}/sample_sheet.csv\"\n",
    "SAMPLESHEET_RNASEQ=\"${OUTDIR_RNASEQ}/sample_sheet.csv\"\n",
    "GENES_OF_INTEREST=\"${OUTDIR_TRIMBA}/genes_of_interest.txt\"\n",
    "MOTIFS_FILE=\"${OUTDIR_TRIMBA}/motif_file.txt\"\n",
    "PARAMS_TRIMBA=\"${OUTDIR_TRIMBA}/nf-params.yml\"\n",
    "CONFIG_TRIMBA=\"${OUTDIR_TRIMBA}/nextflow.config\"\n",
    "PARAMS_RNASEQ=\"${OUTDIR_RNASEQ}/nf-params.yml\"\n",
    "CONFIGRNASEQ=\"${OUTDIR_RNASEQ}/nextflow.config\"\n",
    "\n",
    "#####\n",
    "#\n",
    "# TRIMBA_SAMPLESHEET - Edit this\n",
    "#\n",
    "# This is the main input file for the `trimba` pipeline\n",
    "# If no riboseq data is available, only provide the header line.\n",
    "cat > \"${SAMPLESHEET_TRIMBA}\" <<EOFSST\n",
    "sample,fastq_1,fastq_2,strandedness\n",
    "Riboseq_Control,\"${FASTQ_DIR}/RIBO_DEMO/Ribo-Seq_of_MEF_Cell_Control.fastq.gz\",,unstranded\n",
    "Riboseq_Starvation,\"${FASTQ_DIR}/RIBO_DEMO/Ribo-Seq_of_MEF_Cell_Amino_Acid_Starvation.fastq.gz\",,unstranded\n",
    "QTIseq_Control,\"${FASTQ_DIR}/RIBO_DEMO/QTI-Seq_of_MEF_Cell_Control.fastq.gz\",,unstranded\n",
    "QTIseq_Starvation,\"${FASTQ_DIR}/RIBO_DEMO/QTI-Seq_of_MEF_Cell_Amino_Acid_Starvation.fastq.gz\",,unstranded\n",
    "EOFSST\n",
    "\n",
    "#####\n",
    "#\n",
    "# RNASEQ_SAMPLESHEET - Edit this\n",
    "#\n",
    "# This contains the the input files for the `nf-core/rnaseq` pipeline.\n",
    "# You should provide any orignal RNASEQ data and should also use the data in the `bowtie_rna` directory\n",
    "# The latter will be created by the `trimba` pipeline, but you can anticipate what the name will be\n",
    "cat > \"${SAMPLESHEET_RNASEQ}\" <<EOFSSR\n",
    "sample,fastq_1,fastq_2,strandedness\n",
    "RNASeq_Control,\"${FASTQ_DIR}/RIBO_DEMO/RNA-Seq_of_MEF_Cell_Control.fastq.gz\",,auto\n",
    "RNASeq_Starvation,\"${FASTQ_DIR}/RIBO_DEMO/RNA-Seq_of_MEF_Cell_Amino_Acid_Starvation.fastq.gz\",,auto\n",
    "Riboseq_Control,\"${OUTDIR_TRIMBA}/bowtie_rna/Riboseq_Control.unmapped.fastq.gz\",,auto\n",
    "Riboseq_Starvation,\"${OUTDIR_TRIMBA}/bowtie_rna/Riboseq_Starvation.unmapped.fastq.gz\",,auto\n",
    "QTIseq_Control,\"${OUTDIR_TRIMBA}/bowtie_rna/QTIseq_Control.unmapped.fastq.gz\",,auto\n",
    "QTIseq_Starvation,\"${OUTDIR_TRIMBA}/bowtie_rna/QTIseq_Starvation.unmapped.fastq.gz\",,auto\n",
    "EOFSSR\n",
    "\n",
    "#####\n",
    "#\n",
    "# GENES_OF_INTEREST - Edit this\n",
    "#\n",
    "# This contains the genes of interest for this `trimba` run\n",
    "cat > \"${GENES_OF_INTEREST}\" <<EOFGOI\n",
    "Brca1\n",
    "Zfand2a\n",
    "Tcea1\n",
    "Gm7341\n",
    "EOFGOI\n",
    "\n",
    "#####\n",
    "#\n",
    "# MOTIFS_FILE - Edit this\n",
    "#\n",
    "# This contains the list of MOTIF files to be used in the known-motif part of the `trima` pipeline\n",
    "# Note that you cannot mix DNA and RNA motif files, but for all RNA files there are DNA \"equivalents\"\n",
    "cat > \"${MOTIFS_FILE}\" <<EOFMTFS\n",
    "${MEME_DB}/CISBP-RNA/Mus_musculus.dna_encoded.meme\n",
    "${MEME_DB}/CISBP-RNA/Homo_sapiens.dna_encoded.meme\n",
    "${MEME_DB}/RNA/Ray2013_rbp_Mus_musculus.dna_encoded.meme\n",
    "${MEME_DB}/RNA/Ray2013_rbp_Homo_sapiens.dna_encoded.meme\n",
    "${MEME_DB}/MOUSE/HOCOMOCOv11_core_MOUSE_mono_meme_format.meme\n",
    "${MEME_DB}/MOUSE/HOCOMOCOv11_full_MOUSE_mono_meme_format.meme\n",
    "${MEME_DB}/MOUSE/HOCOMOCOv10_MOUSE_mono_meme_format.meme\n",
    "${MEME_DB}/MOUSE/chen2008.meme\n",
    "${MEME_DB}/HUMAN/HOCOMOCOv11_core_HUMAN_mono_meme_format.meme\n",
    "${MEME_DB}/HUMAN/HOCOMOCOv11_full_HUMAN_mono_meme_format.meme\n",
    "${MEME_DB}/HUMAN/HOCOMOCOv10_HUMAN_mono_meme_format.meme\n",
    "${MEME_DB}/HUMAN/HOCOMOCOv9.meme\n",
    "${MEME_DB}/MIRBASE/22/Homo_sapiens_hsa.dna_encoded.meme\n",
    "${MEME_DB}/MIRBASE/22/Mus_musculus_mmu.dna_encoded.meme\n",
    "EOFMTFS\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8628477a-aa48-4022-9ecc-eae4c6309711",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Pipeline parameters - Review and edit this (Occasionally)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "def6ef83-4e2d-45c9-99d1-7efddd422893",
   "metadata": {},
   "source": [
    "The next set of configurations change the behaviour of the tools in the pipeline. Review the settings and determine if they are still appropriate for the pipeline run, but it is unlikely they need changing between pipeline runs.\n",
    "\n",
    "### `trimba` Pipeline parameter definitions\n",
    "\n",
    "|Parameter|Description|Notes|\n",
    "|-|-|-|\n",
    "|**input**|Path to the SAMPLESHEET|**Required**|\n",
    "|**outdir**|Path to the folder where the results will be stored|**Required**|\n",
    "|**fasta**|Path to the Genome FASTA file|**Required**|\n",
    "|**gtf**|Path to the Genome GTF file|**Required**|\n",
    "|**gtfdb**|Path the the database file for the GTF file|Optional. If `save_reference` is used, can be found in **\\[WFDIR\\]/genome/\\[GENOME\\].gtf.db**|\n",
    "|**rna_index**|Path the the `bowtie` index for the RNA sequences|Optional. If `save_reference` is used, can be found in **\\[WFDIR\\]/genome/RNA/bowtie/**|\n",
    "|**bowtie_index**|Path to the `bowtie` index for the Genome sequences|Optional. If `save_reference` is used, can be found in **\\[WFDIR\\]/genome/GENOME/bowtie**|\n",
    "|**transcripts_index**|Path to the `bowtie` index for the Transcriptome sequences.|Optional. If `save_reference` is used, can be found in **\\[WFDIR\\]/genome/TRANSCRIPTOME/bowtie**|\n",
    "|||\n",
    "|**genes_file**|Path to the file of genes of interest|Optional. Use Gensymbol, one per line|\n",
    "|**motifs_file**|Path to the file of MEME database files|Optional. If not provided, no _known_ motif search will be done, but still will find _novel_ motifs|\n",
    "|||\n",
    "|**skip_trimming**|Boolean, indicating if trimming should be skipped|Note: Booleans should have the value `true` or `false`, case sensitive|\n",
    "|**skip_riboseq**|Boolean, indicating if Riboseq data is provided| Use this, if only providing list of genes. Note: Booleans are case sensitive, `true` or `false`|\n",
    "|**trimmer**|Set this to the trimming program to use|One of `cutadapt`, `trimgalore` or `fastp`. Default is `cutadapt`|\n",
    "|**extra_cutadapt_args**|Extra parameters to give to `cutadapt`||\n",
    "|**rnas_to_filter**|List of GTF feature types to use to filter the reads on|Default \"scRNA,3prime_overlapping_ncRNA,miRNA,snRNA,macro_lncRNA, sRNA,lincRNA,Mt_rRNA,scaRNA,snoRNA, rRNA,Mt_tRNA,bidirectional_promoter_lncRNA,misc_RNA\"|\n",
    "|**nmotifs**|Number of _novel_ motifs to find in the 5'-UTR|Default = 1|\n",
    "|**extra_align_rna_args**|Extra arguments to provide to `bowtie` when aligning reads to _RNA_ for removal||\n",
    "|**extra_align_genome_args**|Extra arguments to provide to `bowtie` when aligning reads to the _genome_|Usually should be '-m 1'|\n",
    "|**extra_align_transcriptome_args**|Extra arguments to provide to `bowtie` when aligning reads to the _transcriptome_|Usually should be '-m 1'|\n",
    "|**extra_ribotish_qual_args**|Extra arguments to provide to `ribotish quality` tool||\n",
    "|**extra_ribotish_args**|Extra arguments to provide to `ribotish quality` tool||\n",
    "\n",
    "\n",
    "### `nf-core/rnaseq` Pipeline parameter definitions\n",
    "\n",
    "These are _some_ of the parameters for the nf-core/rnaseq pipeline that are important. For a full list, see: https://nf-co.re/rnaseq/3.14.0/parameters\n",
    "|Parameter|Description|Notes|\n",
    "|-|-|-|\n",
    "|**input**|Path to the SAMPLESHEET|**Required**|\n",
    "|**outdir**|Path to the folder where the results will be stored|**Required**|\n",
    "|**fasta**|Path to the Genome FASTA file|**Required**|\n",
    "|**gtf**|Path to the Genome GTF file|**Required**|\n",
    "|**salmon_index**|Path the the `salmon` index for the transcript sequences|Optional. If `save_reference` is used, can be found in **\\[WFDIR\\]/genome/index/salmon/**|\n",
    "|**pseudo_aligner**|String indicating which pseudo alginer to use|Default is \"salmon\"|\n",
    "\n",
    "\n",
    "```bash\n",
    "#####\n",
    "#\n",
    "# TRIMBA Pipeline parameters - Review and edit this (occasionally)\n",
    "#\n",
    "# These are the settings for the tools in the pipeline\n",
    "# Review these, but you unlikley need to change these for different pipeline runs\n",
    "cat > \"${PARAMS_TRIMBA}\" <<EOFPARAMST\n",
    "input:                          \"$SAMPLESHEET_TRIMBA\"\n",
    "outdir:                         \"$OUTDIR_TRIMBA\"\n",
    "fasta:                          \"$GRCM38_FASTA\"\n",
    "gtf:                            \"$GRCM38_GTF\"\n",
    "gtfdb:                          \"$GRCM38_GTF_DB\"\n",
    "rna_index:                      \"$GRCM38_RNAS_BOWTIE_IDX\"\n",
    "bowtie_index:                   \"$GRCM38_BOWTIE_IDX\"\n",
    "transcripts_index:              \"$GRCM38_TRANSCRIPTS_BOWTIE_IDX\"\n",
    "\n",
    "genes_file:                     \"$GENES_OF_INTEREST\"\n",
    "motifs_file:                    \"$MOTIFS_FILE\"\n",
    "\n",
    "skip_trimming:                  true\n",
    "skip_riboseq:                   false\n",
    "trimmer:                        \"cutadapt\"\n",
    "extra_cutadapt_args:            '--trim-n --match-read-wildcards -u 16 -n 4 -a AGATCGGAAGAGCACACGTCTG -a AAAAAAAA -a GAACTCCAGTCAC -e 0.2 --nextseq-trim 20 -m 17 -M 34'\n",
    "rnas_to_filter:                 \"scRNA,3prime_overlapping_ncRNA,miRNA,snRNA,macro_lncRNA,sRNA,lincRNA,Mt_rRNA,scaRNA,snoRNA,rRNA,Mt_tRNA,bidirectional_promoter_lncRNA,misc_RNA\"\n",
    "nmotifs:                        1\n",
    "extra_align_rna_args:           ''\n",
    "extra_align_genome_args:        '-m 1'\n",
    "extra_align_transcriptome_args: '-m 1'\n",
    "extra_ribotish_qual_args:       ''\n",
    "extra_ribotish_args:            ''\n",
    "\n",
    "save_reference:                 false\n",
    "EOFPARAMST\n",
    "\n",
    "#####\n",
    "#\n",
    "# RNASEQ Pipeline parameters - Review and edit this (occasionally)\n",
    "#\n",
    "# These are the settings for the tools in the pipeline\n",
    "# Review these, but you unlikley need to change these for different pipeline runs\n",
    "cat > \"${PARAMS_RNASEQ}\" <<EOFPARAMSR\n",
    "input:                 \"$SAMPLESHEET_RNASEQ\"\n",
    "outdir:                \"$OUTDIR_RNASEQ\"\n",
    "fasta:                 \"${GRCM38_FASTA}\"\n",
    "gtf:                   \"${GRCM38_GTF}\"\n",
    "salmon_index:          \"${GRCM38_SALMON_INDEX}\"\n",
    "pseudo_aligner:        \"salmon\"\n",
    "EOFPARAMSR\n",
    "\n",
    "#####\n",
    "#\n",
    "# RNASEQ Extra options - Review and edit this (occasionally)\n",
    "#\n",
    "# The RNASEQ pipeline can be configured with MANY MORE options, and see the main nf-core/rnaseq website for options\n",
    "# https://nf-co.re/rnaseq/3.14.0\n",
    "# To run the most mimimal (and fastest) RNASEQ pipeline, you can skip a lot of the steps by providing these SKIP options\n",
    "FAST=\"--skip_gtf_filter --skip_gtf_transcript_filter --skip_bbsplit --skip_umi_extract --skip_trimming --skip_alignment \\\n",
    "      --skip_markduplicates --skip_bigwig --skip_stringtie --skip_fastqc --skip_preseq --skip_dupradar \\\n",
    "      --skip_qualimap --skip_rseqc --skip_biotype_qc --skip_deseq2_qc --skip_multiqc --skip_qc\"\n",
    "      \n",
    "#************************************************************************************************************************\n",
    "#* END OF CONFIGURATION                                                                                                 *\n",
    "#************************************************************************************************************************\n",
    "\n",
    "#####\n",
    "#\n",
    "# RUN TRIMBA pipeline\n",
    "#\n",
    "cd $OUTDIR_TRIMBA\n",
    "$NEXTFLOW run $TRIMBA -profile docker -resume \\\n",
    "  -params-file \"${PARAMS_TRIMBA}\"\n",
    "  \n",
    "#####\n",
    "#\n",
    "# RUN RNASEQ pipeline\n",
    "#\n",
    "cd $OUTDIR_RNASEQ\n",
    "$NEXTFLOW run nf-core/rnaseq -revision 3.14.0 -profile docker -resume \\\n",
    "  -params-file \"${PARAMS_RNASEQ}\" \\\n",
    "  $FAST\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba78efb7-599f-4a95-a079-f13d4c9658d8",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Run the `trimba` and `nf-core/rnaseq` pipelines"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26703667-013c-4ddf-9bd8-a4cc685a952f",
   "metadata": {},
   "source": [
    "Once the script has been edited and reviewed, you can run the script and it will exectute the two pipelines (one after the other, since `nf-core/rnaseq` requires the output of `trimba` to be available).\n",
    "\n",
    "You can run it in the foreground if you want, but running it in the background would be safer, since it would continue to run, even if you close down the terminal (although not usre how this works on WSL2 ubuntu on Windows)\n",
    "\n",
    "#### Run in the background\n",
    "```bash\n",
    "nohup run_trimba_and_rnaseq.sh > pipe.out 2> pipe.err &\n",
    "```\n",
    "\n",
    "#### Run in the foreground\n",
    "\n",
    "```bash\n",
    "./run_trimba_and_rnaseq.sh\n",
    "```\n",
    "\n",
    "All the log files are in the output directory (they are hidden, so you need to use `ls -al` to see them)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46c49dfb-9e7b-4af2-9e4a-e23d2d58cd81",
   "metadata": {},
   "source": [
    "# Runnning `trimba` and `nf-core/rnaseq` for the first time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9178fcf-bf7f-49d9-bc74-b46a055a8690",
   "metadata": {},
   "source": [
    "The first time you run the pipelines, you can use it to have it create the indices for the RNA and DNA database and various other indices it requires.\n",
    "\n",
    "I created a seperate script called `first_run_trimba_and_rnaseq.sh`.\n",
    "\n",
    "You still need to review the top part to ensure the various directories are appropriate for you but then you can simply run and it will create the directories `trimba_firstrun` and `rnaseq_firstrun` and you can extract the various indices and place them in the appropriate place on your ubuntu server.\n",
    "\n",
    "**See the table above with the location of the various indices**\n",
    "\n",
    "It will run `trimba` without any Riboseq data and will run the demo for `nf-core/rnaseq`.\n",
    "\n",
    "```bash\n",
    "./run_trimba_and_rnaseq.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "234c145e-ffd5-42bb-82e0-67b191efda2e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
